Makefile

```diff
@@ -0,0 +1,39 @@
+.PHONY = default deps build test start clean start-database
+
+IMAGE_NAME := codelytv/typescript-ddd-skeleton
+SERVICE_NAME := app
+
+# Test if the dependencies we need to run this Makefile are installed
+DOCKER := $(shell command -v docker)
+DOCKER_COMPOSE := $(shell command -v docker-compose)
+deps:
+ifndef DOCKER
+	@echo "Docker is not available. Please install docker"
+	@exit 1
+endif
+ifndef DOCKER_COMPOSE
+	@echo "docker-compose is not available. Please install docker-compose"
+	@exit 1
+endif
+
+default: build
+
+# Build image
+build:
+	docker build  -t $(IMAGE_NAME):dev .
+
+# Run tests
+test: build
+	docker-compose run --rm $(SERVICE_NAME) bash -c 'npm run build && npm run test'
+
+# Start the application
+start: build
+	docker-compose up $(SERVICE_NAME) && docker-compose down
+
+# Clean containers
+clean:
+	docker-compose down --rmi local --volumes --remove-orphans
+
+# Start mongodb container in background
+start_database:
+	docker-compose up -d mongo
```

Comment:
What do you think about adding the Node dependencies management to the Makefile in order to make it easier to handle them from within the Node Docker container?

You have an [example in the PHP repository](https://github.com/CodelyTV/php-ddd-example/blob/master/Makefile#L13-L23).

In that case, we even create a temporal Docker container to run Composer (the PHP dependencies manager). This way we don't need to install Composer in the PHP container, but because Node integrates it all in the same service, we could simplify it just directly interacting with the Node Container.

The goal behind this would be to ensure we install/update dependencies using the very same Node version that executes our code in an easy way (just a `make`) command 😊

Reply:
Good idea! We will take a look and include that possibility here.

Reply:
Not sure if I am missing something, but I think we are getting the requirement of ensuring that we install the dependencies using the same Node version every time we launch a `make` command. We are getting that by setting the `build` step as a dependency for all the rules.

Example:

```bash
# Build image
build:
	docker build  -t $(IMAGE_NAME):dev .

# Run tests
test: build
	docker-compose run --rm $(SERVICE_NAME) bash -c 'npm run build && npm run test'
```

Setting the build as a dependency of the `test` step we have a tradeoff: the test step is building the image every time it's launched, but it is ensuring that the dependencies are correct. Currently, it is not taking much time once the docker image was built for the first time so we will leave as it is for now. If this is slow in the future, we can create an `install` rule in the Makefile as you said.